### PR TITLE
feat: llama index data loader construct

### PR DESCRIPTION
**Pull request by scottschreckengaust**
This construct is to load data via [LlamaIndex](https://www.llamaindex.ai/)

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+import { NagSuppressions } from 'cdk-nag';
+import { Construct } from 'constructs';
+import { BaseClass, BaseClassProps, ConstructName } from '../../../common/base-class';
+
+export interface LlamaIndexDataLoaderProps {
+  /**
+   * The directory to build the Docker image
+   * @description The directory to build the Docker image.
+   */
+  readonly dockerImageAssetDirectory: string;
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:35]: as discussed here we can add a default dockerfile if none is provided

Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts]: done

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+   * @default - _dev
+   */
+  readonly stage?: string;
+  /**
+   * Enable observability. Warning: associated cost with the services
+   * used. Best practive to enable by default.
+   *
+   * @default - true
+   */
+  readonly observability?: boolean;
```
[the remark this answers is not in the document]
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:70]: Added VPC option.

File: README.md [hunk 3]
```diff
@@ -139,6 +139,8 @@ The following constructs are available in the library:
 | [Content Generation](./src/patterns/gen-ai/aws-contentgen-appsync-lambda/README.md) | Generate images from text using Amazon titan-image-generator-v1 or stability.stable-diffusion-xl-v1 model.  | AWS Lambda, Amazon Bedrock, AWS AppSync |
 | [Web crawler](./src/patterns/gen-ai/aws-web-crawler/README.md) | Crawl websites and RSS feeds on a schedule and store changeset data in an Amazon Simple Storage Service bucket.  | AWS Lambda, AWS Batch, AWS Fargate, Amazon DynamoDB  |
 | [Amazon Bedrock Monitoring (Amazon CloudWatch Dashboard)](./src/patterns/gen-ai/aws-bedrock-cw-dashboard/README.md) | Amazon CloudWatch dashboard to monitor model usage from Amazon Bedrock. | Amazon CloudWatch |
+| [TXT to SQL](./src/patterns/gen-ai/aws-bedrock-cw-dashboard/README.md) | Amazon CloudWatch dashboard to monitor model usage from Amazon Bedrock. | Amazon CloudWatch |
```
Comment (krokoko) [inline README.md:142]: description for text to sql seems wrong ?

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+| existingDashboard               | [aws_cloudwatch.Dashboard](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudwatch.Dashboard.html)                                | ![Optional](https://img.shields.io/badge/optional-4169E1) | Existing dashboard to be used by the construct. **Mutually exclusive** with ```dashboardName``` - only one should be specified.                                                                                                                                                                                                                                                           |
+| dashboardName | string | ![Optional](https://img.shields.io/badge/optional-4169E1) | A name for the dashboard which will be created. If not provided, the construct will create a new dashboard named 'AossMetricsDashboard'. **Mutually exclusive** with ```existingDashboard``` - only one should be specified.                                                                                                                                                                                                                                                            |
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/README.md:81]: pattern construct props don't match the construct

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+## Pattern Properties
+
+| **Name**                     | **Type**                                                                                                                  | **Description**                                                                                                                                                                |
+| :----------------------------- | :-------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| dashboard                          | [aws_cloudwatch.Dashboard](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudwatch.Dashboard.html)                                     | The CloudWatch Dashboard used by the construct (whether created by the construct or provided by the client)                                                                                     |
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/README.md:87]: same as above for this section

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+## Methods
+
+N/A
+
+## Default properties
+
+Out-of-the-box implementation of the construct without any override will set the following defaults:
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/README.md:95]: Need to provide the default properties (memory limit,...) without any override from the user

File: resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
+#!/bin/env python3 -x -u
```
Comment (krokoko) [inline resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py:1]: header license file

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+      enforceSSL: true,
+      deadLetterQueue: {
+        maxReceiveCount: 3,
+        queue: new Queue(this, 'DeadLetterQueue', {
+          enforceSSL: true,
+        }),
+      },
+    });
+    const queueProcessingFargateService = new QueueProcessingFargateService(this, 'Service', {
+      cluster: new Cluster(this, 'Cluster', {
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:178]: Documentation for https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs_patterns.QueueProcessingFargateService.html#cluster mentions that if a cluster is provided (which seems to always be the case here), the vpc should be omitted. vpc is here optional, so what happens if a user passes a vpc through the props ? 
If a vpc is passed by the user, do we need to add any specific info to the doc about the vpc passed ?

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+        containerInsights: true,
+      }),
+      vpc: props.vpc,
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:181]: If the vpc is passed by props or managed by the QueueProcessingFargateService, shouldn't it have a vpc endpoint for S3 ?

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+      });
+    }
+    bucketsInvolved.push(this.outputBucket);
+    // Create a new SSM Parameter holding a String
+    const circuitBreakerParameter = new StringParameter(this, 'CircuitBreaker', {
+      stringValue: 'False',
+    });
+    const asset = new DockerImageAsset(this, 'Image', {
+      directory: this.dockerImageAssetDirectory,
+      platform: Platform.LINUX_AMD64,
```
Comment (krokoko) [inline src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:165]: why are we setting this to Linux AMD 64 ?

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts (hunk 9 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:181]: The default creation, when `undefined`, does not create the S3 VPC endpoint. I can document the security practice to use VPC endpoints for the VPC

Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:181]: The default creation, when `undefined`, does not create the S3 VPC endpoint. I can document the security practice to use VPC endpoints for the VPC

File: resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py (hunk 7 above)
Comment (scottschreckengaust) [reply resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py:1]: fixed 270965b5accba2b07509cab3dd213ef6bd4a509a

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md (hunk 6 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/README.md:95]: fixed 97a94356b0cd6484195e73760c92c740b30c57c9

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md (hunk 4 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/README.md:81]: Sorry had some local fixes that I needed to push 97a94356b0cd6484195e73760c92c740b30c57c9

File: src/patterns/gen-ai/aws-llama-index-data-loader/README.md (hunk 5 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/README.md:87]: corrected 97a94356b0cd6484195e73760c92c740b30c57c9

File: README.md (hunk 3 above)
Comment (scottschreckengaust) [reply README.md:142]: opps, this snuck in...fixing this...

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts (hunk 10 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:165]: Due to macOS, the base image doesn't have a tag for the platform like the lambda does, so defining that explicitely so alternative libraries that are pulled in from Docker's build do not mix in for python. This version has fixed to x86_64 to reduce issues such as https://github.com/aws/aws-cdk/issues/30962

File: README.md (hunk 3 above)
Comment (scottschreckengaust) [reply README.md]: good catch!

File: README.md [hunk 11]
```diff
@@ -139,6 +139,8 @@ The following constructs are available in the library:
...
+| [LlamaIndex Data Loading](./src/patterns/gen-ai/aws-bedrock-cw-dashboard/README.md) | Amazon CloudWatch dashboard to monitor model usage from Amazon Bedrock. | Amazon CloudWatch |
```
[the remark this answers is not in the document]
Comment (scottschreckengaust) [reply README.md]: fixed

File: src/patterns/gen-ai/aws-llama-index-data-loader/index.ts (hunk 8 above)
Comment (scottschreckengaust) [reply src/patterns/gen-ai/aws-llama-index-data-loader/index.ts:178]: So the task running can access what it needs. I'll have to work out what those boundaries are. Should this be in a followup PR?

**Review by krokoko — APPROVED** [review]

File: resources/gen-ai/aws-llama-index-data-loader/docker/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,34 @@
+ARG PYTHON_TAG="latest"
```
Comment (dineshSajwan) [inline resources/gen-ai/aws-llama-index-data-loader/docker/Dockerfile:1]: Missing License header

File: resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+def save_document(documents, serialize=True) -> None:
+    save_result = False
+    for document in documents:
+        file_name = f"{document.metadata['file_name']}" if 'file_name' in document.metadata else '' 
+        LOGGER.info(f'document id: {document.doc_id}')
+        LOGGER.debug(f'document text: {document.text}')
+        LOGGER.debug(f'document metadata: {document.metadata}')
+        LOGGER.debug(f'document extra info: {document.extra_info}')
+        # print(f'INFO(save_document): document id: {document.doc_id}', flush=True)
+        # print(f'DEBUG(save_document): document text: {document.text}', flush=True)
```
Comment (dineshSajwan) [inline resources/gen-ai/aws-llama-index-data-loader/docker/sqs_consumer.py:94]: please removed the commented print statements

File: src/common/base-class/construct-name-enum.ts [hunk 14]
```diff
@@ -21,5 +21,6 @@ export enum ConstructName {
   AWSCONTENTGENAPPSYNCLAMBDA = 'C8',
   AWSRAGAPPSYNCSTEPFNKENDRA = 'C9',
   AWSWEBCRAWLER = 'C10',
-  AWSTEXTTOSQL= 'C11'
+  AWSTEXTTOSQL= 'C11',
+  AWSLLAMAINDEXDATALOADER = 'C12',
```
Comment (dineshSajwan) [inline src/common/base-class/construct-name-enum.ts:25]: please add AWSLLAMAINDEXDATALOADER in `protected static usageMetricMap : Record<string, number> ` in common/base-class.ts for generating metrics

**Review by dineshSajwan — COMMENTED** [review]
nit picks

**Review by krokoko — APPROVED** [review]

File: src/common/base-class/construct-name-enum.ts (hunk 14 above)
Comment (scottschreckengaust) [reply src/common/base-class/construct-name-enum.ts:25]: done

File: resources/gen-ai/aws-llama-index-data-loader/docker/Dockerfile (hunk 12 above)
Comment (scottschreckengaust) [reply resources/gen-ai/aws-llama-index-data-loader/docker/Dockerfile:13]: added

**Review by krokoko — APPROVED** [review]

**Review by dineshSajwan — APPROVED** [review]
LGTM

**Pull request merged**